File: android/src/main/java/dev/locus/core/EventDispatcher.java
package dev.locus.core;

import android.os.Handler;
import android.os.Looper;

import java.util.Map;

import io.flutter.plugin.common.EventChannel;

public class EventDispatcher {
    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    private final HeadlessDispatcher headlessDispatcher;
    private EventChannel.EventSink eventSink;

    public EventDispatcher(HeadlessDispatcher headlessDispatcher) {
        this.headlessDispatcher = headlessDispatcher;
    }

    public void setEventSink(EventChannel.EventSink eventSink) {
        this.eventSink = eventSink;
    }

    public void sendEvent(Map<String, Object> event) {
        if (eventSink == null) {
            headlessDispatcher.dispatch(event);
            return;
        }
        mainHandler.post(() -> {
            if (eventSink != null) {
                eventSink.success(event);
            }
        });
    }
}

File: android/src/main/kotlin/dev/locus/core/EventDispatcher.kt
package dev.locus.core

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel

class EventDispatcher(
    private val headlessDispatcher: HeadlessDispatcher
) {
    private val mainHandler = Handler(Looper.getMainLooper())

    // `@Volatile` so writes from the Flutter platform thread (when the
    // engine attaches/detaches) are visible to readers on background
    // threads (sync executors, headless dispatch path). Without it, the
    // Java memory model permits a stale-cache `null` read after attach
    // — events would silently fall through to the headless dispatcher.
    @Volatile
    private var eventSink: EventChannel.EventSink? = null

    fun setEventSink(sink: EventChannel.EventSink?) {
        eventSink = sink
    }

    fun sendEvent(event: Map<String, Any>) {
        val sink = eventSink
        if (sink == null) {
            headlessDispatcher.dispatch(event)
            return
        }
        mainHandler.post {
            sink.success(event)
        }
    }
}
